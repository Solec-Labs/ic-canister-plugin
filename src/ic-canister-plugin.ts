import { EnvironmentPlugin, type Compiler } from 'webpack';
import {
  CanisterId,
  CanisterName,
  DfxCanisterConfig,
  getCanisterConfig,
} from './dfx-canister-config';
import { getDfxNetwork } from './dfx-network';
import type { IcWebpackPluginLogger } from './ic-webpack-plugin-logger';

type CanisterDefinitions = Record<CanisterName, CanisterId>;

export class IcCanisterPlugin {
  public apply(compiler: Compiler, logger: IcWebpackPluginLogger): void {
    const network = getDfxNetwork(logger);
    const canisterConfig = getCanisterConfig(network);
    const canisterDefinitions = this.getCanisterDefinitions(
      logger,
      canisterConfig,
      network,
    );

    new EnvironmentPlugin(canisterDefinitions).apply(compiler);
  }

  private getCanisterDefinitions(
    logger: IcWebpackPluginLogger,
    canisterConfig: DfxCanisterConfig,
    network: string,
  ): CanisterDefinitions {
    return Object.entries(canisterConfig).reduce<CanisterDefinitions>(
      (accum, curr) => {
        const [canisterName, canisterDetails] = curr;
        const canisterEnvironmentName =
          canisterName.toUpperCase() + '_CANISTER_ID';
        const canisterId = canisterDetails[network] ?? '';

        logger.log(
          `Set ${canisterName} environment variable ${canisterEnvironmentName} to ${canisterId}`,
        );

        return {
          ...accum,
          [canisterEnvironmentName]: canisterId,
        };
      },
      {},
    );
  }
}
