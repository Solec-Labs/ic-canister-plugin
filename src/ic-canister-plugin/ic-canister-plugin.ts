import { EnvironmentPlugin, type Compiler } from 'webpack';
import {
  CanisterId,
  CanisterName,
  CanisterConfig,
  DfxNetwork,
  getCanisterConfig,
  getDfxNetwork,
  IcWebpackPluginLogger,
} from '../common';

type CanisterDefinitions = Record<CanisterName, CanisterId>;

export class IcCanisterPlugin {
  public apply(compiler: Compiler): void {
    const logger = compiler.getInfrastructureLogger('IcCanisterPlugin');

    const dfxNetwork = getDfxNetwork(logger);
    const canisterConfig = getCanisterConfig(dfxNetwork);
    const canisterDefinitions = this.getCanisterDefinitions(
      canisterConfig,
      dfxNetwork,
      logger,
    );

    new EnvironmentPlugin(canisterDefinitions).apply(compiler);
  }

  private getCanisterDefinitions(
    canisterConfig: CanisterConfig,
    dfxNetwork: DfxNetwork,
    logger: IcWebpackPluginLogger,
  ): CanisterDefinitions {
    return Object.entries(canisterConfig).reduce<CanisterDefinitions>(
      (accum, curr) => {
        const [canisterName, canisterDetails] = curr;
        const canisterEnvironmentName =
          canisterName.toUpperCase() + '_CANISTER_ID';
        const canisterId = canisterDetails[dfxNetwork] ?? '';

        logger.log(
          `Set ${canisterName} environment variable ${canisterEnvironmentName} to ${canisterId}`,
        );

        return {
          ...accum,
          [canisterEnvironmentName]: canisterId,
        };
      },
      {},
    );
  }
}
