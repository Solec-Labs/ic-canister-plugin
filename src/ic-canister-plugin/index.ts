export * from './ic-canister-plugin';
