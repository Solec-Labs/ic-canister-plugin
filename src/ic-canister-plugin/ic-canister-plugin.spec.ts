import * as fs from 'fs';
import * as webpack from 'webpack';
import { createMock } from 'ts-auto-mock';
import { On, method } from 'ts-auto-mock/extension';
import type { Compiler } from 'webpack';
import type { CanisterConfig, IcWebpackPluginLogger } from '../common';
import { IcCanisterPlugin } from './ic-canister-plugin';

jest.mock('fs');

jest.mock('webpack', () => {
  const realWebpack = jest.requireActual<typeof webpack>('webpack');

  return {
    WebpackError: realWebpack.WebpackError,
    EnvironmentPlugin: jest.fn(() => ({ apply: jest.fn() })),
  };
});

describe('IcCanisterPlugin', () => {
  let plugin: IcCanisterPlugin;

  let loggerMock: IcWebpackPluginLogger;
  let compilerMock: Compiler;
  let readFileSyncMock: jest.SpyInstance;
  let environmentPluginMock: jest.SpyInstance;

  const canisterConfig: CanisterConfig = {
    canister_frontend: {
      local: 'rrkah-fqaaa-aaaaa-aaaaq-cai',
      testnet: 'ryjl3-tyaaa-aaaaa-aaaba-cai',
      ic: 'r7inp-6aaaa-aaaaa-aaabq-cai',
    },
    canister_backend: {
      local: 'rkp4c-7iaaa-aaaaa-aaaca-cai',
      testnet: 'renrk-eyaaa-aaaaa-aaada-cai',
      ic: 'rno2w-sqaaa-aaaaa-aaacq-cai',
    },
  };
  const canisterConfigString = JSON.stringify(canisterConfig);
  const canisterConfigBuffer = Buffer.from(canisterConfigString);

  beforeEach(() => {
    readFileSyncMock = jest.spyOn(fs, 'readFileSync');

    compilerMock = createMock<Compiler>();
    loggerMock = createMock<IcWebpackPluginLogger>({
      log: jest.fn(),
    });
    On(compilerMock)
      .get(method('getInfrastructureLogger'))
      .mockReturnValue(loggerMock);

    environmentPluginMock = jest.spyOn(webpack, 'EnvironmentPlugin');

    plugin = new IcCanisterPlugin();

    delete process.env.DFX_NETWORK;
    delete process.env.NODE_ENV;
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should create an EnviornmentPlugin with the local canister definitions', () => {
    readFileSyncMock.mockReturnValue(canisterConfigBuffer);

    plugin.apply(compilerMock);

    expect(environmentPluginMock).toHaveBeenCalledTimes(1);
    expect(environmentPluginMock.mock.calls[0]).toMatchSnapshot();
  });

  it('should create an EnviornmentPlugin with the mainnet canister definitions', () => {
    readFileSyncMock.mockReturnValue(canisterConfigBuffer);
    process.env.NODE_ENV = 'production';

    plugin.apply(compilerMock);

    expect(environmentPluginMock).toHaveBeenCalledTimes(1);
    expect(environmentPluginMock.mock.calls[0]).toMatchSnapshot();
  });

  it('should create an EnviornmentPlugin with explicitly defined canister definitions', () => {
    readFileSyncMock.mockReturnValue(canisterConfigBuffer);
    process.env.DFX_NETWORK = 'testnet';

    plugin.apply(compilerMock);

    expect(environmentPluginMock).toHaveBeenCalledTimes(1);
    expect(environmentPluginMock.mock.calls[0]).toMatchSnapshot();
  });
});
