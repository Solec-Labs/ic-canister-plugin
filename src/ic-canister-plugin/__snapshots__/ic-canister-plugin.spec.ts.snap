// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IcCanisterPlugin should create an EnviornmentPlugin with explicitly defined canister definitions 1`] = `
[
  {
    "CANISTER_BACKEND_CANISTER_ID": "renrk-eyaaa-aaaaa-aaada-cai",
    "CANISTER_FRONTEND_CANISTER_ID": "ryjl3-tyaaa-aaaaa-aaaba-cai",
  },
]
`;

exports[`IcCanisterPlugin should create an EnviornmentPlugin with the local canister definitions 1`] = `
[
  {
    "CANISTER_BACKEND_CANISTER_ID": "rkp4c-7iaaa-aaaaa-aaaca-cai",
    "CANISTER_FRONTEND_CANISTER_ID": "rrkah-fqaaa-aaaaa-aaaaq-cai",
  },
]
`;

exports[`IcCanisterPlugin should create an EnviornmentPlugin with the mainnet canister definitions 1`] = `
[
  {
    "CANISTER_BACKEND_CANISTER_ID": "rno2w-sqaaa-aaaaa-aaacq-cai",
    "CANISTER_FRONTEND_CANISTER_ID": "r7inp-6aaaa-aaaaa-aaabq-cai",
  },
]
`;
